src/background.c

```diff
@@ -0,0 +1,817 @@
+    {
+      if (data->stamp < stamp)
+        {
+          if (data->handle)
+            close_notification (data->handle);
+          g_hash_table_iter_remove (&iter);
+        }
+    }
+}
+
+static char *
+flatpak_instance_get_display_name (FlatpakInstance *instance)
+{
+  const char *app_id = flatpak_instance_get_app (instance);
+  if (app_id[0] != 0)
+    {
+      g_autofree char *desktop_id = NULL;
+      g_autoptr(GAppInfo) info = NULL;
+
+      desktop_id = g_strconcat (app_id, ".desktop", NULL);
+      info = (GAppInfo*)g_desktop_app_info_new (desktop_id);
+
+      if (info)
+        return g_strdup (g_app_info_get_display_name (info));
+    }
+
+  return g_strdup (app_id);
+}
+
+static void
+kill_instance (const char *id)
+{
+  InstanceData *idata;
+
+  idata = g_hash_table_lookup (applications, id);
+
+  if (idata)
+    {
+      g_debug ("Killing app %s", flatpak_instance_get_app (idata->instance));
+      kill (flatpak_instance_get_child_pid (idata->instance), SIGKILL);
+    }
+}
+
+typedef struct {
+  char *app_id;
+  char *id;
+  Permission perm;
+} DoneData;
+
+static void
+done_data_free (gpointer data)
+{
+  DoneData *ddata = data;
+
+  g_free (ddata->app_id);
+  g_free (ddata->id);
+  g_free (ddata);
+}
+
+typedef enum {
+  FORBID = 0,
+  ALLOW  = 1,
+  IGNORE = 2
+} NotifyResult;
+
+static void
+notify_background_done (GObject *source,
+                        GAsyncResult *res,
+                        gpointer data)
+{
+  DoneData *ddata = (DoneData *)data;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GVariant) results = NULL;
+  guint response;
+  guint result;
+  InstanceData *idata;
+
+  if (!xdp_impl_background_call_notify_background_finish (background_impl,
+                                                          &response,
+                                                          &results,
+                                                          res,
+                                                          &error))
+    {
+      g_warning ("Error from background backend: %s", error->message);
+      done_data_free (ddata);
+      return;
+    }
+
+  g_variant_lookup (results, "result", "u", &result);
+
+  if (result == ALLOW)
+    {
+      g_debug ("Allowing app %s to run in background", ddata->app_id);
+
+      if (ddata->perm != ASK)
+        ddata->perm = YES;
+    }
+  else if (result == FORBID)
+    {
+      g_debug ("Forbid app %s to run in background", ddata->app_id);
+
+      if (ddata->perm != ASK)
+        ddata->perm = NO;
+
+      kill_instance (ddata->id);
+    }
+  else if (result == IGNORE)
+    {
+      g_debug ("Allow this instance of %s to run in background without permission changes", ddata->app_id);
+    }
+  else
+    g_debug ("Unexpected response from NotifyBackground: %u", result);
+
+  if (ddata->perm != UNSET)
+    set_permission (ddata->app_id, ddata->perm);
+
+  idata = g_hash_table_lookup (applications, ddata->id);
```

Comment:
This is accessing the applications hashtable on the main thread, so it will need some locking!